File: src/proyecto_2/proyecto_2.py
# Proyecto 2:
# python3 proyecto_2
# Angel Armando Valay Martinez

from OpenGL.GLUT import *
from OpenGL.GL import *
from OpenGL.GLU import *
import math
import random
import sys
import array as arr
import game
import numpy

PI=3.14159265358979324

# Dimensiones de la pantalla
WidthWin = 600
HeightWin = 600

# variables globales
NEAR_Z =  1.0
FAR_Z = 400.0

# zoom por ejes
zoom = 60
zoom_x = 0
zoom_y = 0
zoom_z = 0

# isShowMessage?
showMessageStart = 0

# Timer
refreshMillis = 500

# Se crea una malla del eje Y
wall1 = game.Wall(3, 3, (-10, -10, 0), 'Y')
listPoints = wall1.getPoints()

# Se crea una malla del eje Z
wall2 = game.Wall(3,10,(-10,-10,0),'Z')
listPoints2 = wall2.getPoints()

# Se crea una malla del eje X
wall3 = game.Wall(10,3,(-10,-10,0),'X')
listPoints3 = wall3.getPoints()

# array de figuras
# este array guarda las figuras que se van generando para asi mantener todo las figuras
figures = []


def timer(value):
    global refreshMillis, figures, showMessageStart
    # Si el juego ya empezo
    if showMessageStart == 1:
        figure = figures[0]
        print('figura -> ', figure)
        for cube in figure.cubes:
            print(cube.getPointToArray())
        if figure.isValidateMove(0,-1,0):
            print('puede bajar')
            # Refresca la pantalla
            figure.moveY(-1)
            glutPostRedisplay()
        else:
            print('ya no se puede bajar')
            # figures=[]
            # Se crea una figura de prueba
            figures.insert(0,game.Figure(random.randint(0,3), 1, 10, 1))
            print(figures)
            showMessageStart = 0
            glutPostRedisplay()
    glutTimerFunc(refreshMillis, timer, 0)

# Para mostrar las opciones
def printInteraction():
    print("Interaction:")
    print("Presiona [ESCAPE] para finalizar.")

# Funcion para iniciar
def init():
    glClearColor(0.0, 0.0, 0.0, 0.0)

# Funcion para redimensionar la pantalla
def resize(w, h):
    if h == 0:
        h = 1
    glViewport(0, 0, w, h)
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    gluPerspective(10.0, w / h, NEAR_Z, FAR_Z)
    glMatrixMode(GL_MODELVIEW)

# Funcion para dibujar texto
def drawTextOnDisplay(string):
    for character in string:
        glutBitmapCharacter(GLUT_BITMAP_HELVETICA_18, ord(character))

# FUncion para el keyinput
def keyInput(key, x, y):
    global zoom, zoom_x, zoom_y, zoom_z, showMessageStart
    if key == b'\053':  # ESCAPE
        zoom = zoom  - 1

    if key == b'\055':  # +
        zoom = zoom + 1

    if key == b'\033':  # -
        zoom = zoom - 1

    if key == b'\170':  # x
        zoom_x = zoom_x + 10

    if key == b'\171':  # y
        zoom_y = zoom_y + 10

    if key == b'\172':  # z
        zoom_z = zoom_z + 10

    if key == b'\040': # space
        if showMessageStart == 0:
            showMessageStart = 1
        elif showMessageStart == 1:
            showMessageStart = 0

    if key == b'\062': # borrar es de prueba 2222
        if figures[-1].isValidateMove(0, 0, 1):
            figures[-1].moveZ(1)
    if key == b'\070': # borrar es de prueba 8888
        if figures[-1].isValidateMove(0, 0, -1):
            figures[-1].moveZ(-1)
    if key == b'\064': # borrar es de prueba 4444
        if figures[-1].isValidateMove(-1, 0, 0):
            figures[-1].moveX(-1)
    if key == b'\066': # borrar es de prueba 6666
        if figures[-1].isValidateMove(1, 0, 0):
            figures[-1].moveX(1)

    glutPostRedisplay()

# Funcion para mostrar los datos
def displayText():
    glColor(1, 1, 1)
    glRasterPos3f(-10.0, 10.0, 0)
    drawTextOnDisplay('Tetris 3D')
    glRasterPos3f(-10.0, 9.0, 0)
    drawTextOnDisplay('Tiempo: 00:00')
    # si aun no empieza o esta en pausa
    if showMessageStart == 0:
        glRasterPos3f(-10.0, 8.0, 0)
        drawTextOnDisplay('Presionar SPACE para empezar o reanudar')

# Display
def displayMe():
    global listPoints, zoom, zoom_x, zoom_y, zoom_z
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
    glMatrixMode(GL_MODELVIEW)
    glLoadIdentity()
    gluLookAt(zoom, zoom, zoom, 0.0, 5, 0.0, 0.0, 1.0, 0.0)
    glRotatef(zoom_x, 1.0, 0.0, 0.0)
    glRotatef(zoom_y, 0.0, 1.0, 0.0)
    glRotatef(zoom_z, 0.0, 0.0, 1.0)
    # Agregar lookAt
    # Mostrar texto
    displayText()
    # Dibuja la primera malla
    glBegin(GL_LINES)
    for item in listPoints:
        glVertex3fv(item)
    glEnd()
    glColor(0.5,0.6,0.7)
    # Dibuja la segunda malla
    glBegin(GL_LINES)
    for item in listPoints2:
        glVertex3fv(item)
    glEnd()
    glColor(0.8, 0.1, 0.7)
    # Dibuja la tercera malla
    glBegin(GL_LINES)
    for item in listPoints3:
        glVertex3fv(item)
    glEnd()
    # hago un translei para mover las figuras una mitad del cubo hacia el inferior para que
    # coloquen bien
    glTranslatef(game.UNIT_CUBE/2,game.UNIT_CUBE/2,game.UNIT_CUBE/2)
    # se imprimen las figuras
    for figure in  figures:
        # muestra la figura cubo x cubo
        glColor(0.6, 0.6, 0.6)
        for cube in figure.cubes:
            # mostrar la cara del cubo cara x cara
            for face in range(1,6,1):
                # glPointSize(5)
                # dibuja el poligono
                glBegin(GL_POLYGON)
                # extrae los puntos
                for point in cube.Faces(face):
                    glVertex3fv(point)
                glEnd()
    glutSwapBuffers()

# Funcion principal
def main():
    # se anade la figura de prueba a la lista
    figures.append(game.Figure(random.randint(0, 3), 1, 10, 1))
    print(figures)
    printInteraction()
    glutInit(sys.argv)
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB)
    glutInitWindowPosition(100, 100)
    glutInitWindowSize(HeightWin, WidthWin)
    glutCreateWindow("Juego 2 | Tetris 3D")
    glutReshapeFunc(resize)
    glutDisplayFunc(displayMe)
    glutKeyboardFunc(keyInput)
    glutTimerFunc(0,timer,0)
    glutMainLoop()

# Ejecutando el main
main()

File: src/proyecto_2/proyecto_2.cpp
//
// 2 JUEGO
// ANGEL ARMANDO VALAY MARTINEZ
// Created by angel on 14/11/19.
//
#include <stdlib.h>
#include <GL/glut.h>
#include <iostream>
#include <math.h>
#include "game.h"

// Globals.
static float square_color[3] = {1.0, 0.0, 0.0}; // Color of the square.


//Dimensiones de la pantalla
#define WidthWin 600
#define HeightWin 600

//variables globales
double NEAR_Z = 1.0;
double FAR_Z = 400.0;

//zoom por ejes
float zoom = 80;
float zoom_x = 0;
float zoom_y = 0;
float zoom_z = 0;

// is show message?
bool showMessageStart = true;

// timer
int refreshMillis = 800;

//Mallas del juego
Wall wallX = Wall(LIM_X,LIM_Z,'Y');
vector<Point> pointsWallX = wallX.getPoints();
Wall wallY = Wall(LIM_X,15,'Z');
vector<Point> pointsWallY = wallY.getPoints();
Wall wallZ = Wall(15,LIM_Z,'X');
vector<Point> pointsWallZ = wallZ.getPoints();

vector<Figure*> figures;


Figure *figure = new Figure(rand() % 3 ,rand() % 1+5,18,rand() % 3+5);

void timer(int value){
    for(Figure *figure1: figures){
        for(Cube  cube: figure1->cubes()) {
           if(cube.getPoints()->y > 16){
               std::cout<<"#################### FIN DEL JUEGO ##############"<<endl;
                exit(0);
           }
        }
    }
    if (showMessageStart)
        if (figure->isValidateMove(0,-1,0, &figures)){
            //refrescar pantalla
            figure->moveY(-1);
            glutPostRedisplay();
        }else{
            figures.push_back(figure);
            figure = new Figure(rand() % 3 ,rand() % 1+5,18,rand() % 1+5);
            glutPostRedisplay();
        }
    glutTimerFunc(refreshMillis, timer, 0);
}

void restartGame(){
    showMessageStart = false;
    for(Figure* figure1: figures){
        delete(figure1);
    }
    figures.clear();
    figure = new Figure(rand() % 3 ,rand() % 1+5,18,rand() % 1+5);
    glutPostRedisplay();
}

void printInteraction(){
    cout<<"Interation:"<<endl;
    cout<<"Key [p] for stop/resume"<<endl;
    cout<<"Key [z] for rotate on axis Z"<<endl;
    cout<<"Key [x] for rotate on axis X"<<endl;
    cout<<"Key [c] for rotate on axis Y"<<endl;
    cout<<"Key [UP ARROW] for move to far"<<endl;
    cout<<"Key [DOWN ARROW] for move to front"<<endl;
    cout<<"Key [LEFT ARROW] for move to left"<<endl;
    cout<<"Key [RIGHT ARROW] for move to right"<<endl;
    cout<<"Key [+] for more velocity"<<endl;
    cout<<"Key [END] for restart game"<<endl;
}

void resize(int w, int h){
    if (h == 0){
        h = 1;
    }
    glViewport(0, 0, w, h);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(10.0, w / h, NEAR_Z, FAR_Z);
    glMatrixMode(GL_MODELVIEW);
}

void key(unsigned char key, int x, int y){
    switch (key){
        case 'p':
            showMessageStart = !showMessageStart;
            break;
        case 'x':
            if (figure->isValidateRotateX(&figures)){
                figure->rotationX();
            }
            break;
        case 'c':
            if (figure->isValidateRotateY(&figures)){
                figure->rotationY();
            }
            break;
        case 'z':
            if (figure->isValidateRotateZ(&figures)){
                figure->rotationZ();
            }
            break;
        case 'u':
            zoom_y+=10;
//            cout<<"y "<<zoom_y<<endl;
            break;
        case 'i':
            zoom_x+=10;
//            cout<<"x "<<zoom_x<<endl;
            break;
        case 'o':
            zoom_z+=10;
//            cout<<" "<<zoom_z<<endl;
            break;
        case '+':
            if (showMessageStart)
                if (figure->isValidateMove(0,-1,0, &figures)){
                    //refrescar pantalla
                    figure->moveY(-1);
                    glutPostRedisplay();
                }else{
                    figures.push_back(figure);
                    figure = new Figure(rand() % 3,3,18,3);
                    glutPostRedisplay();
                }
            break;
        default:
            break;
    }
    if (zoom_x > 360)
        zoom_x = 10;
    if (zoom_x < 0)
        zoom_x = 360;
    if (zoom_y > 360)
        zoom_y = 10;
    if (zoom_y < 0)
        zoom_y = 360;
    glutPostRedisplay();
}

// The top menu callback function.
void top_menu(int id)
{
    if (id==1) exit(0);
}
void sub_menu(int id){
    glutPostRedisplay();
}

// The sub-menu callback function.
void view_menu(int id)
{
    if (id==2)
    {
        zoom_x = 200;
        zoom_y =240;
        zoom_z =310;
    }

    if (id==4)
    {
        zoom_x = 0;
        zoom_y =0;
        zoom_z = 0;
    }
    glutPostRedisplay();
}
// Routine to make the menu.
void makeMenu(void)
{
    // The sub-menu is created first (because it should be visible when the top
    // menu is created): its callback function is registered and menu entries added.
    int sub_menu;
    sub_menu = glutCreateMenu(view_menu);
    glutAddMenuEntry("TOP", 2);
    glutAddMenuEntry("NORMAL",4);

    // The top menu is created: its callback function is registered and menu entries,
    // including a submenu, added.
    glutCreateMenu(top_menu);
    glutAddSubMenu("View", sub_menu);
    glutAddMenuEntry("Quit",1);

    // The menu is attached to a mouse button.
    glutAttachMenu(GLUT_RIGHT_BUTTON);
}


void displayMe(){
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    gluLookAt(zoom, zoom, zoom,0.0,5,0.0,0.0,1.0,0.0);
    glRotatef(zoom_x,1,0,0);
    glRotatef(zoom_y,0,1,0);
    glRotatef(zoom_z,0,0,1);
    //display text
    glBegin(GL_LINES);
    for(Point &point: pointsWallX){
        glVertex3d(point.x,point.y, point.z);
    }
    glEnd();
    glColor3f(0.5,0.6,0.7);
    glBegin(GL_LINES);
    for(Point &point: pointsWallY){
        glVertex3d(point.x,point.y, point.z);
    }
    glEnd();
    glColor3f(0.8, 0.1, 0.7);
    glBegin(GL_LINES);
    for(Point &point: pointsWallZ){
        glVertex3d(point.x,point.y, point.z);
    }
    glEnd();
    //hago un translei para mover las figuras una mitad del cubo hacia el inferior para que
    // coloquen bien
    glTranslatef(UNIT_CUBE/2,UNIT_CUBE/2,UNIT_CUBE/2);
//    se imprimen las figuras almacenadas
    for(Figure* figure1 : figures){
        glColor3fv(figure1->getColor());
        for (Cube cube: figure1->cubes()){
            for (int i = 0; i < 6; ++i) {
                glBegin(GL_POLYGON);
                for (int j = 0; j < 4; ++j) {
                    glVertex3d(cube.getFaces(i)[j].x, cube.getFaces(i)[j].y, cube.getFaces(i)[j].z);
                }
                glEnd();
            }
        }
        glColor3f(1,1,1);
        for (Cube cube: figure1->cubes()){
            for (int i = 0; i < 6; ++i) {
                glBegin(GL_LINE_LOOP);
                for (int j = 0; j < 4; ++j) {
                    glVertex3d(cube.getFaces(i)[j].x, cube.getFaces(i)[j].y, cube.getFaces(i)[j].z);
                }
                glEnd();
            }
        }
    }
    //aqui se imprimen la figura seleccionada
    glColor3fv(figure->getColor());
    for (Cube cube: figure->cubes()){
        for (int i = 0; i < 6; ++i) {
            glBegin(GL_POLYGON);
            for (int j = 0; j < 4; ++j) {
                glVertex3d(cube.getFaces(i)[j].x, cube.getFaces(i)[j].y, cube.getFaces(i)[j].z);
            }
            glEnd();

        }
    }
    glColor3f(1,1,1);
    for (Cube cube: figure->cubes()){
        for (int i = 0; i < 6; ++i) {
            glBegin(GL_LINE_LOOP);
            for (int j = 0; j < 4; ++j) {
                glVertex3d(cube.getFaces(i)[j].x, cube.getFaces(i)[j].y, cube.getFaces(i)[j].z);
            }
            glEnd();

        }
    }
    glutSwapBuffers();
}

void specialKeyInput(int key, int x, int y){
    if (key == GLUT_KEY_UP)
    {
        if (figure->isValidateMove(0,0,-1, &figures)){
            figure->moveZ(-1);
            glutPostRedisplay();
        }
    }
    if (key == GLUT_KEY_DOWN)
    {
        if (figure->isValidateMove(0,0,1, &figures)){
            figure->moveZ(1);
            glutPostRedisplay();
        }
    }
    if (key == GLUT_KEY_RIGHT)
    {
        if (figure->isValidateMove(1,0,0, &figures)){
            figure->moveX(1);
            glutPostRedisplay();
        }
    }
    if (key == GLUT_KEY_LEFT)
    {
        if (figure->isValidateMove(-1,0,0, &figures)){
            figure->moveX(-1);
            glutPostRedisplay();
        }
    }
    if(key == GLUT_KEY_END){
        restartGame();
    }
}

void init(){
    glClearColor(0.0, 0.0, 0.0, 0.0);
    // Make menu.
    makeMenu();
}

int main(int argc, char **argv){
    printInteraction();
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_RGB | GLUT_DOUBLE | GLUT_DEPTH | GLUT_STENCIL | GLUT_MULTISAMPLE);
    glutInitWindowSize(HeightWin, WidthWin);
    glutInitWindowPosition(200, 200);
    glutCreateWindow("Juego 2 | Tetris 3D");
    glutReshapeFunc(resize);
    glutDisplayFunc(displayMe);
    glutKeyboardFunc(key);
    glutTimerFunc(0,timer,0);
    glutSpecialFunc(specialKeyInput);
    init();
    glutMainLoop();
    return 0;
}